networkx/algorithms/isomorphism/tree_isomorphism.py

```diff
@@ -1,29 +1,1021 @@
-"""
-An algorithm for finding if two undirected trees are isomorphic,
-and if so returns an isomorphism between the two sets of nodes.
+"""An algorithm for finding if two undirected trees are isomorphic. 
 
 This algorithm uses a routine to tell if two rooted trees (trees with a
 specified root node) are isomorphic, which may be independently useful.
 
-This implements an algorithm from:
-The Design and Analysis of Computer Algorithms
-by Aho, Hopcroft, and Ullman
-Addison-Wesley Publishing 1974
-Example 3.2 pp. 84-86.
-
-A more understandable version of this algorithm is described in:
-Homework Assignment 5
-McGill University SOCS 308-250B, Winter 2002
-by Matthew Suderman
-http://crypto.cs.mcgill.ca/~crepeau/CS250/2004/HW5+.pdf
+References
+----------
+  .. [1] A. V. Aho, J. E. Hopcroft y J. D. Ullman, "The Desing And Analysis of
+     Computer Algorithms", Addison Wesley Publishing Company, (1974): 78-86.
+
+  .. [2] M. Suderman, "Homework Assignment 5", McGill University SOCS 308-250B,
+     (Winter 2002): http://crypto.cs.mcgill.ca/~crepeau/CS250/2004/HW5+.pdf
+
+  .. [3] G. Valiente, "Algorithms on Trees and Graphs", Springer (2002): 
+     151-166.
+
+  .. [4] M. Carrasco-Ruiz, "Herramientas para el Manejo Computacional de
+     Gráficas", Bachelor of Science Thesis, UNAM, (2022): 55-77.
+
 """
 
+from collections import Counter
+
 import networkx as nx
 from networkx.utils.decorators import not_implemented_for
 
 __all__ = ["rooted_tree_isomorphism", "tree_isomorphism"]
 
 
+def categorize_entries(S):
+    """Groups the entries of the lists by its position.
+
+    Given a list of lists whose entries are numbers between 0 and m. Define a
+    NONEMPTY map such that NONEMPTY(i) are the numbers in the i-th position of
+    some list in S; this numbers are sorted.
+
+    For example, given ((1), (1, 2, 3), (1, 2, 2)), NONEMPTY(2) = [2, 3].
+
+    Parameters
+    ----------
+    S : list of lists of ints (natural numbers)
+        The list of lists of natural numbers.
+
+    Returns
+    -------
+    NONEMPTY : dict of int: list of ints
+        The NONEMPTY map such that NONEMPTY(i) are the numbers in the i-th
+        position of some list in S; this list of numbers is sorted.
+
+    Notes
+    -----
+    This algorithm runs in O(l_total + m) time and O(l_total) space, where
+    l_total is the sum of the lengths of the lists in S and m is the max value
+    of all the lists in S.
+
+    """
+    # Define the an empty NONEMPTY map.
+    NONEMPTY = {}
+
+    # The max length found in S.
+    l_max = max([len(s) for s in S])
```

Comment:
Minor nitpick : you don't need the square brackets in this line (since `max` works on iterators)